src/server.jsx

```diff
@@ -37,9 +56,31 @@ export default async function ({ req, res, context }) {
     return res.redirect(context.url);
   }
 
-  // console.log('###', context);
+  const promises = routes.reduce((matches, route) => {
+    const match = matchPath(req.url, route);
+    if (match) {
+      matches.push(
+        route.component.fetchData ?
+          route.component.fetchData({ match, dispatch: store.dispatch }) : Promise.resolve(null),
+      );
+    }
+    return matches;
+  }, []);
+
+  await Promise.all(promises);
+
+  // We need to wait for the promise to complete to dispatch the actions
+  const markup = ReactDOMServer.renderToString(
+    <Provider store={store}>
+      <Router location={req.url} context={{}} ><AppView /></Router>
+    </Provider>,
+  );
```

Comment:
The bad part of this, is that we need to render the app twice in the server, once to check if the user is allowed in the route, otherwise redirect, and then after the requests are made and store is updated, we generate the real code sent to the client